### ISSUE TITLE
Bypass.expect should be able to match on multiple routes

### ISSUE THREAD

**Issue by ericmj** [PSPDFKit-labs/bypass#10]
Example:

``` elixir
Bypass.expect(bypass, "GET", "/user/123", fn conn ->
  # return user 123
end)
Bypass.expect(bypass, "GET", "/user/124", fn conn ->
  # return user 124
end)
```

**Comment by MSch** [comment]
Hi Eric! I found your fork, do you mind if I pull in those commits?

https://github.com/PSPDFKit-labs/bypass/compare/master...ericmj:emj-multi-bypass

**Comment by ericmj** [comment]
It needs tests and cleaned up, especially the support for multiple bypasses. But I do believe it is all working, at least it works in our tests :).

**Comment by gaynetdinov** [comment]
Do you have any plans to bring this functionality anytime soon?

**Comment by alco** [comment]
@gaynetdinov It is easy to implement this by matching on the `conn` argument as can be seen [here](https://github.com/hexpm/hex/blob/17988c72db64ee1383247326eb7d444e7c107c57/test/mix/tasks/hex/public_keys_test.exs#L172-L180):

```elixir
Bypass.expect bypass, fn
  %Plug.Conn{request_path: "/registry.ets.gz"} = conn ->
    ...
  %Plug.Conn{request_path: "/registry.ets.gz.signed"} = conn ->
    ...
end
```

It may even be easier to use `%Plug.Conn{:path_info}`. You can also match on `%Plug.Conn{:method}`.

@ericmj I noticed that Hex doesn't use the functionality implemented in your fork of Bypass yet it is still references it in `mix.exs`.

**Comment by ericmj** [comment]
I think Hex uses some other functionality from my fork. This specific
functionality is used in a private project.

On Wednesday, 1 June 2016, Alexei Sholik <notifications@github.com> wrote:

> @gaynetdinov <https://github.com/gaynetdinov> It is easy to implement
> this by matching on the conn argument as can be seen here
> <https://github.com/hexpm/hex/blob/17988c72db64ee1383247326eb7d444e7c107c57/test/mix/tasks/hex/public_keys_test.exs#L172-L180>
> :
>
> Bypass.expect bypass, fn
>   %Plug.Conn{request_path: "/registry.ets.gz"} = conn ->
>     ...
>   %Plug.Conn{request_path: "/registry.ets.gz.signed"} = conn ->
>     ...end
>
> It may even be easier to use %Plug.Conn{:path_info}. You can also match
> on %Plug.Conn{:method}.
>
> @ericmj <https://github.com/ericmj> I noticed that Hex doesn't use the
> functionality implemented in your fork of Bypass yet it is still references
> it in mix.exs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PSPDFKit-labs/bypass/issues/10#issuecomment-223073551>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AATV2pEDV-ZcEdKse1d-rYYMpbx6faq7ks5qHceLgaJpZM4HhsJJ>
> .
>


-- 
Eric Meadows-Jönsson

**Comment by gaynetdinov** [comment]
@alco thanks!

**Comment by ericmj** [comment]
@alco Hex uses this commit from my fork: https://github.com/PSPDFKit-labs/bypass/commit/92f949a5dacbce6b11a76df2600fc7b2076c8bd8 because needs to be built and tested for all Elixir versions. It has been merged upstream so I can switch to this master.

**Comment by pollingj** [comment]
I have an issue where I need to his the same request_path twice, but want a different response.  Clearly the below doesn't work, just wondering if anyone could suggest how I could get it to work?

This is a for a SOAP request.

```
Bypass.expect bypass, fn
  %Plug.Conn{request_path: "/"} = conn ->
    ...
  %Plug.Conn{request_path: "/"} = conn ->
    ...
end
```

**Comment by alco** [comment]
@pollingj you an achieve this by saving some indicator in ETS or an Agent after the first request. Then when the second request arrives, you'll know to return a different response.

**Comment by MSch** [comment]
@ericmj we just released bypass 0.7.0 with the changes from https://github.com/PSPDFKit-labs/bypass/pull/36 - please let me know if this fixes your issue

**Comment by ericmj** [comment]
Awesome, I will try it out and let you know if I have any issues.
